### PR TITLE
Create Parent Proposal and use gov-created client in Handshake

### PR DESCRIPTION
**Pull request by AdityaSripal**
closes: https://github.com/cosmos/ibc-go/issues/392
supercedes: https://github.com/cosmos/ibc-go/pull/439

- Creates a proposal type and handling functions for creating a new child chain
- Upon passing the proposal, the IBC client for the child chain is created automatically
- The CCV channel on the parent chain for that child chain **MUST** be built on top of the above client.
- This ensures that the channel is speaking to the correct child chain and not an imposter

### CODE REVIEW

File: x/ccv/types/keys.go [hunk 1]
```diff
@@ -10,6 +10,8 @@ const (
 
 	// ChannelStatusKeyPrefix is the key prefix for storing the validation status of the CCV channel
 	ChannelStatusKeyPrefix = "channelstatus"
+
+	RouterKey = "ccv"
```
Comment (colin-axner) [inline x/ccv/types/keys.go:14]: should this use the ccv module name instead?

File: x/ccv/types/errors.go [hunk 2]
```diff
@@ -18,4 +18,6 @@ var (
 	ErrInvalidUnbondingSequence = sdkerrors.Register(ModuleName, 11, "invalid unbonding sequence")
 	ErrInvalidUnbondingTime     = sdkerrors.Register(ModuleName, 12, "child chain has invalid unbonding time")
 	ErrInvalidChildState        = sdkerrors.Register(ModuleName, 13, "parent chain has invalid state for child chain")
+	ErrInvalidChildClient       = sdkerrors.Register(ModuleName, 14, "ccv channel is not built on correct client")
```
Comment (colin-axner) [inline x/ccv/types/errors.go:21]: should this indicate what a correct client might look like?

File: x/ccv/parent/proposal_handler.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+// NewCreateChildChainHandler defines the CCV parent proposal handler
+func NewCreateChildChainHandler(k keeper.Keeper) govtypes.Handler {
+	return func(ctx sdk.Context, content govtypes.Content) error {
+		switch c := content.(type) {
+		case *ccv.CreateChildChainProposal:
+			return k.CreateChildChainProposal(ctx, c)
+		default:
+			return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized ibc proposal content type: %T", c)
```
Comment (colin-axner) [inline x/ccv/parent/proposal_handler.go:18]:
```suggestion
			return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized ccv proposal content type: %T", c)
```

File: x/ccv/parent/keeper/keeper.go [hunk 4]
```diff
@@ -188,11 +188,14 @@ func (k Keeper) VerifyChildChain(ctx sdk.Context, channelID string) error {
...
 	}
-	_, tmClient, err := k.getUnderlyingClient(ctx, channelID)
+	clientID, tmClient, err := k.getUnderlyingClient(ctx, channelID)
 	if err != nil {
 		return err
 	}
-	// TODO: Verify consensus state against initial validator set of the child chain
+	ccvClientId := k.GetChildClient(ctx, tmClient.ChainId)
+	if ccvClientId != clientID {
+		return sdkerrors.Wrapf(ccv.ErrInvalidChildClient, "CCV channel must be built on top of CCV client: %s, got %s", ccvClientId, clientID)
```
Comment (colin-axner) [inline x/ccv/parent/keeper/keeper.go:197]:
```suggestion
		return sdkerrors.Wrapf(ccv.ErrInvalidChildClient, "CCV channel must be built on top of CCV client: expected %s, got %s", ccvClientId, clientID)
```

File: x/ccv/parent/keeper/genesis_test.go [hunk 5]
```diff
@@ -11,9 +11,9 @@ func (suite *KeeperTestSuite) TestGenesis() {
 	// set some chain-channel pairs before exporting
 	ctx := suite.parentChain.GetContext()
 	for i := 0; i < 4; i++ {
-		suite.parentChain.App.(*app.App).ParentKeeper.SetChainToChannel(ctx, fmt.Sprintf("chainid-%d", i), fmt.Sprintf("channel-%d", i))
-		suite.parentChain.App.(*app.App).ParentKeeper.SetChannelToChain(ctx, fmt.Sprintf("channel-%d", i), fmt.Sprintf("chainid-%d", i))
-		suite.parentChain.App.(*app.App).ParentKeeper.SetChannelStatus(ctx, fmt.Sprintf("channel-%d", i), types.Status(i))
+		suite.parentChain.App.(*app.App).ParentKeeper.SetChainToChannel(ctx, fmt.Sprintf("chainid-%d", i), fmt.Sprintf("channelid-%d", i))
```
Comment (colin-axner) [inline x/ccv/parent/keeper/genesis_test.go:14]: Should you be referencing a variable/const instead of hardcoding?

File: x/ccv/types/proposal.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	clienttypes "github.com/cosmos/ibc-go/modules/core/02-client/types"
+	"github.com/cosmos/ibc-go/modules/core/exported"
+)
+
+const (
+	ProposalTypeCreateChildChain = "CreateChildChain"
+)
+
+var (
+	_ govtypes.Content = &CreateChildChainProposal{}
```
Comment (colin-axner) [inline x/ccv/types/proposal.go:20]: add the UnpackInterfaces var assertion?

**Review by colin-axner — COMMENTED** [review]
Left various suggestions/nits, reviewed most files but not all

File: proto/interchain_security/ccv/v1/ccv.proto [hunk 7]
```diff
@@ -30,3 +32,28 @@ enum Status {
...
+  string title = 1;
+  // the description of the proposal
+  string description = 2;
+  // the proposed chain-id of the new child chain, must be different from all other child chain ids of the executing
+  // parent chain.
+  string chain_id = 3 [(gogoproto.moretags) = "yaml:\"chain_id\""];
+  // the proposed client state of new child chain.
+  // Since it contains no consensus information, this may be created before child chain is created.
+  google.protobuf.Any client_state = 4 [(gogoproto.moretags) = "yaml:\"client_state\""];
+  // genesis hash with no staking information included.
```
Comment (fedekunze) [inline proto/interchain_security/ccv/v1/ccv.proto:53]: > with no staking information included

i.e without the genesis module included? We should also specify in the comment that this is the child chain's genesis hash

**Review by jtremback — COMMENTED** [review]
LGTM

**Pull request merged**